Allow forcing HTTP Connect proxy passthrough for Konnectivity support

steeling opened the pull request:
Konnectivity server only accepts HTTP CONNECT passthrough proxying. The standard lib proxies via a GET request for destinations that are non-TLS however. This means that when scraping a non-secured HTTP endpoint through a proxy, the request is always sent as a GET. Only HTTPS destinations are sent as a CONNECT Passthrough. There is nothing in any spec that requires this however. 

This PR adds a flag to scrape configs, to force HTTP destinations to use CONNECT passthrough.

See #5109

zekker6 commented on lib/promscrape/client.go:
```diff
@@ -95,7 +95,9 @@ func newClient(ctx context.Context, sw *ScrapeWork) (*client, error) {
 	proxyURL := sw.ProxyURL
-	if !isTLS && proxyURL.IsHTTPOrHTTPS() {
+	// If it's not TLS, and the Proxy URL is HTTP(s), and the user didn't request to force HTTP CONNECT Passthrough. The
+	// latter check is to support konnectivity, which only uses HTTP CONNECT passthrough, even for non-TLS requests. #5109
```
Could you please place a full link to an issue here?
It is usually easier to follow the link later on.

steeling replied:
done!

zekker6 commented on lib/promscrape/scrapework.go:
```diff
@@ -107,6 +107,8 @@ type ScrapeWork struct {
 	// ProxyURL HTTP proxy url
 	ProxyURL *proxy.URL
 
+	ProxyForceHTTPConnect bool
```
Is this variable needed here or only one defined in `ProxyClientConfig` is needed?

steeling replied:
So `ScrapeConfig` has a `ProxyClientConfig` which is a natural place to put this. However that config gets translated to a generic struct `promauth.Config` on both `ScrapeWorkConfig` and `ScrapeWork`.

The `promauth.Config` is used for both destination endpoints and proxying, and is just TLS info. Adding it there would leave a dangling appendage on the regular `AuthConfig` field.

That said, I'm open to doing whatever the VM team recommends :)

zekker6 replied:
Thank you for the details, that makes sense to me.

Could you please add a comment for this parameter to match the code style?

steeling replied:
done!

zekker6 reviewed (commented):
Hello @steeling, thank you for the PR!

Please, see a few comments.

Also, to get this merged it is required to update the SD section of documentation here - https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/sd_configs.md
Could you please add description of new parameter there?

steeling commented:
> Hello @steeling, thank you for the PR!
> 
> Please, see a few comments.
> 
> Also, to get this merged it is required to update the SD section of documentation here - https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/sd_configs.md Could you please add description of new parameter there?

done, thanks for the feedback!

zekker6 commented on docs/sd_configs.md:
```diff
@@ -1609,6 +1609,10 @@ and in the majority of [supported service discovery configs](#supported-service-
 
+    # proxy_force_http_connect is an optional boolean to force HTTP bound scrape requests to use HTTP CONNECT Passthrough for proxying.
+    # By default, HTTP CONNECT Passthrough is only used on HTTPS destinations, and HTTP destinations proxy as a GET request.
+      # proxy_force_http_connect: <boolean>
```
```suggestion
    # proxy_force_http_connect: <boolean>
```

steeling replied:
done!

zekker6 commented on lib/promscrape/client_test.go:
```diff
@@ -0,0 +1,82 @@
+
+func TestReadData_HTTPConnectPassthrough(t *testing.T) {
+	ctx := context.Background()
+	trequire := require.New(t)
```
Could you please remove `testify` dependency?
We're usually writing tests with standard Go functions and just making comparisons manually, you can refer to these tests for example: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/promscrape/config_test.go

steeling replied:
done!

zekker6 reviewed (commented):
See a few comments, LGTM other than that.

steeling commented:
@zekker6 PTAL again, in testing this I noticed I hadn't implemented the fix for configs that set `stream_parse` to true. I've implemented this now, and added some more robust tests.

steeling commented:
Hi all, friendly ping on this.

cc: @zekker6, @hagen1778, @f41gh7

hagen1778 commented:
@zekker6 I have a little knowledge of this. @f41gh7 could you please take a look? I remember you did something similar already.

zekker6 commented:
@steeling Sorry for the delay, I've completely missed this. Will take a look either later today or on Monday.

zekker6 reviewed (approved):
LGTM

steeling commented:
thanks @zekker6

Pushed fixes for lint, and signed all commits

steeling commented:
@zekker6, can you help get reviews from remaining approvers?

cc: @hagen1778 , @f41gh7

steeling commented:
Hey folks, if I continued on this PR could we get this in?

zekker6 commented:
Hey @steeling, let me ping other maintainers internally to get the approvals.

> Hey folks, if I continued on this PR could we get this in?

I don't see why not, so it would be great if you could update the PR to get this merged.

hagen1778 commented:
The feature looks good to me. But I need technical approvement from @f41gh7 for implementation.

f41gh7 commented on lib/promscrape/client.go:
```diff
@@ -121,11 +126,15 @@ func newClient(ctx context.Context, sw *ScrapeWork) (*client, error) {
 		setFasthttpProxyHeaders = func(req *fasthttp.Request) error {
 			return proxyURLOrig.SetFasthttpHeaders(sw.ProxyAuthConfig, req)
 		}
+		if pu := proxyURL.GetURL(); pu != nil {
```
for instance, looks like proxyURL = "socks5://some-proxy" could fail. It'd be great to have test for it. But it maybe too complicated.

f41gh7 commented on lib/promscrape/client.go:
```diff
@@ -144,10 +153,16 @@ func newClient(ctx context.Context, sw *ScrapeWork) (*client, error) {
-	if pu := sw.ProxyURL.GetURL(); pu != nil {
-		proxyURLFunc = http.ProxyURL(pu)
+
+	streamDialFunc, err := newStatDialFunc(proxyURL, sw.ProxyAuthConfig, getStreamDialer())
```
Looks like it's possible to use  single `dialFunc` instead of `streamDialFunc`.

f41gh7 reviewed (commented):
Thanks overall looks good. Have a few comments.

Also, it'd be great to have a changelog entry with proposed changes.

f41gh7 commented:
Sorry, I just realized that given PR is outdated. 

There was a major change into master branch, that remove fasthttp client and it makes impossible to merge current PR with it.  And it will requires a significant amount of work and changes to add "CONNECT method" to current branch.

 I see the following options:
1) address issue to the https://github.com/kubernetes-sigs/apiserver-network-proxy . Their proxy server is incompatible with go standard library http/transport. 
2) try to rewrite patch to the vmagent. It'll require changes to the dial func https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/netutil/statdial.go#L15 and https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/promscrape/client.go#L64


 Lets continue communication at linked issue.
